### ISSUE TITLE
Support Older CUDA GPU Hardware

### ISSUE THREAD

**Issue by gel** [karpathy/llm.c#92]
This project is phenomenal, thank you so much for this.

I think the project is currently aiming for:
(1) Education and simplicity - making it simple to understand and build (avoid bloat).
(2) Efficient performance for experimentation for research - making it useful.

I would also consider adding a new target for

(3) Supporting older cheap, globally available strong hardware such as older GPUs - to lower the barrier for experimentation and research.

Therefore, I would try to support by default CUDA10 or CUDA9 and not starting from CUDA11 which prevents many of the older generation GPUs to run.

Concretely, I would love to have replacement for cooperative_gorups/reduce.h which was introduced only in CUDA11 - https://github.com/karpathy/llm.c/commit/2c811981013d95a6712580ffc7b616c2595fa542#diff-bf6b442957e5458cf8baab2a18039fdde86d74199a0864a79e7288fe55f31a98R15

**Comment by dagelf** [comment]
Are you saying it worked on Cuda 10 before that? Also this might be related #237

**Comment by ademeure** [comment]
I agree in principle but I’m confused what are the “many of the older generations of GPUs” you are referring to? NVIDIA still supports the latest CUDA 12.4 with Maxwell which was released in 2014, more than 10 years ago.

The fastest pre-Maxwell GPU is the Kepler GK110 which at ~4TFLOPs is much better than nothing (about the level of the latest iPhone) but for most LLM research you’re likely to be just as limited by its performance as its lack of support :( Do you have anything in mind?

I’d be very strongly in favour of making sure that it keeps working for Maxwell (SM 5.0) on CUDA 12 going forward though which shouldn’t be too difficult.

**Comment by gel** [comment]
> Are you saying it worked on Cuda 10 before that? Also this might be related #237

Before all the latest changes I was successful training using Nvidia Titan X (Pascal) on Cuda 10 installed (before all the latest changes)  - compute level 6.1 (GPU memory = ~12 GBs) on Ubuntu 20.04.
In any case, it might be too opportunistic, but I am very curious if it's possible using all the latest advancements in quantization and training methods to build a very cheap setup from old GPUs and efficient code and train decent models.
